site/en/articles/renderingng-fragmentation/index.md

```diff
@@ -0,0 +1,204 @@
+
+### Legacy engine interaction
+
+Legacy data structures are still in charge of JavaScript APIs that read layout information, so we need to write back data to the legacy engine in a manner that it understands. This includes updating the legacy multi-column data structures, like [LayoutMultiColumnFlowThread](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/layout/layout_multi_column_flow_thread.h;l=144;drc=237a52e7e2b6d22eb1b5912e134384254d5b8974), correctly. 
+
+### Legacy engine fallback detection and handling
+
+We had to fall back to the legacy layout engine when there was content inside that couldn't yet be handled by LayoutNG block fragmentation. At the time of shipping core LayoutNG block fragmentation (spring 2022), that included flex, grid, tables, and anything that's printed. This was particularly tricky because we need to detect the need for legacy fallback before creating [objects in the layout tree](/articles/layoutng/#a-30000-foot-view-of-layout-engine-architectures). For example, before we know if there's a multi-column container ancestor, and before we know which DOM nodes will become a [formatting context](https://developer.mozilla.org/docs/Web/Guide/CSS/Block_formatting_context) or not. It's a chicken-and-egg problem that doesn't have a perfect solution, but as long as its only misbehavior is false positives (fallback to legacy when there's actually no need), it's ok, because any bugs in that layout behavior are ones Chromium already has, not new ones.
+
+### Pre-paint tree walk
+
+That's something we [do after layout](/articles/renderingng-architecture/#rendering-pipeline-structure), but before painting. The main challenge is that we still need to walk the [layout object tree](/articles/layoutng/#a-30000-foot-view-of-layout-engine-architectures), but we have NG fragments now—so how do we deal with that? We walk both the layout object and NG fragment trees at the same time! This is quite complicated, because mapping between the two trees isn't trivial. While the layout object tree structure closely resembles that of the DOM tree, the fragment tree is an **output** of layout, not an input to it. Apart from actually reflecting the effect of any fragmentation including inline fragmentation (line fragments) and block fragmentation (column or page fragments), the fragment tree also has a direct parent-child relationship between a [containing block](https://developer.mozilla.org/docs/Web/CSS/Containing_block) and the DOM descendants that have that fragment as their containing block. For example, in the fragment tree, a fragment generated by an absolutely positioned element is a direct child of its containing block fragment, even if there are other nodes in the ancestry chain between the out-of-flow positioned descendant and its containing block.  
```

Comment:
```suggestion
Pre-paint is something we [do after layout](/articles/renderingng-architecture/#rendering-pipeline-structure), but before painting. The main challenge is that we still need to walk the [layout object tree](/articles/layoutng/#a-30000-foot-view-of-layout-engine-architectures), but we have NG fragments now—so how do we deal with that? We walk both the layout object and NG fragment trees at the same time! This is quite complicated, because mapping between the two trees isn't trivial. While the layout object tree structure closely resembles that of the DOM tree, the fragment tree is an **output** of layout, not an input to it. Apart from actually reflecting the effect of any fragmentation including inline fragmentation (line fragments) and block fragmentation (column or page fragments), the fragment tree also has a direct parent-child relationship between a [containing block](https://developer.mozilla.org/docs/Web/CSS/Containing_block) and the DOM descendants that have that fragment as their containing block. For example, in the fragment tree, a fragment generated by an absolutely positioned element is a direct child of its containing block fragment, even if there are other nodes in the ancestry chain between the out-of-flow positioned descendant and its containing block.  
```